packages/wrangler/src/deployment-bundle/bindings.ts

```diff
@@ -62,3 +69,263 @@ export function getBindings(
 				},
 	};
 }
+
+export type Settings = {
+	bindings: Array<WorkerMetadataBinding>;
+};
+
+type PendingResourceOperations = {
+	create: (name: string) => Promise<string>;
+	updateId: (id: string) => void;
+};
+type PendingResources = {
+	kv_namespaces: (CfKvNamespace & PendingResourceOperations)[];
+	r2_buckets: (CfR2Bucket & PendingResourceOperations)[];
+	d1_databases: (CfD1Database & PendingResourceOperations)[];
+};
+export async function provisionBindings(
+	bindings: CfWorkerInit["bindings"],
+	accountId: string,
+	scriptName: string,
+	autoCreate: boolean
+): Promise<void> {
+	const pendingResources: PendingResources = {
+		d1_databases: [],
+		r2_buckets: [],
+		kv_namespaces: [],
+	};
+	let settings: Settings | undefined;
+
+	try {
+		settings = await getSettings(accountId, scriptName);
+	} catch (error) {
+		logger.debug("No settings found");
+	}
+
+	for (const kv of bindings.kv_namespaces ?? []) {
+		if (!kv.id) {
+			if (inBindingSettings(settings, "kv_namespace", kv.binding)) {
+				kv.id = INHERIT_SYMBOL;
+			} else {
+				pendingResources.kv_namespaces?.push({
+					binding: kv.binding,
+					async create(title) {
+						const id = await createKVNamespace(accountId, title);
+						kv.id = id;
+						return id;
+					},
+					updateId(id) {
+						kv.id = id;
+					},
+				});
+			}
+		}
+	}
+
+	for (const r2 of bindings.r2_buckets ?? []) {
+		if (!r2.bucket_name) {
+			if (inBindingSettings(settings, "r2_bucket", r2.binding)) {
+				r2.bucket_name = INHERIT_SYMBOL;
+			} else {
+				pendingResources.r2_buckets?.push({
+					binding: r2.binding,
+					async create(bucketName) {
+						await createR2Bucket(accountId, bucketName);
+						r2.bucket_name = bucketName;
+						return bucketName;
+					},
+					updateId(bucketName) {
+						r2.bucket_name = bucketName;
+					},
+				});
+			}
+		}
+	}
+
+	for (const d1 of bindings.d1_databases ?? []) {
+		if (!d1.database_id) {
+			if (inBindingSettings(settings, "d1", d1.binding)) {
+				d1.database_id = INHERIT_SYMBOL;
+			} else {
+				pendingResources.d1_databases?.push({
+					binding: d1.binding,
+					async create(name) {
+						const db = await createD1Database(accountId, name);
+						d1.database_id = db.uuid;
+						return db.uuid;
+					},
+					updateId(id) {
+						d1.database_id = id;
+					},
+				});
+			}
+		}
+	}
+
+	if (Object.values(pendingResources).some((v) => v && v.length > 0)) {
```

Comment:
Can we gate this around a service env check and throw otherwise? i.e. `if(!isLegacyEnv(config)){ throw new UserError("You can't provision with a service environment")}`

Reply:
[ee0a46d](https://github.com/cloudflare/workers-sdk/pull/7427/commits/ee0a46d06e4e767bea362478165237ff531d0f69)